### PR TITLE
feat: Save last patched app

### PR DESCRIPTION
**Pull request by BenjaminHalko**
closes: #643 

Adds patch history: It will keep the last 3 APKs of the patched packages. The user can then install or export it.

The UI could use a bit of work though. Right now if you click on the app in the history menu, it will let you install or export. If you open it in the installed app list you get the option to open / uninstall the app like before.

The reason for the 2 separate menus is for the case where someone patches the app, then closes manager without installing. It would be nice to be able to go back and install it. Thus it needs a separate entry from the version that is currently installed on the system.

The history is limited to the 3 most recent packages because it has to keep the patched apk for every entry which uses up space.

Example where Twitch was compiled but not installed, and where Repair Toolkit was compiled and installed:

https://github.com/ReVanced/revanced-manager/assets/73490201/ce041dac-41f7-41e0-a9f7-9b67d85216f0

### CODE REVIEW

**Comment by Ushie** [comment]
You may be interested in adding chips to switch between those cards, look at older versions of revanced manager before the removal of the "update" WIP card

**Comment by validcube** [comment]
Now we need "delete stored APK" in settings!!!!

**Comment by oSumAtrIX** [comment]
This feature requires enhanced transparency. Users should be able to create a backup of the patched app once the patching process is complete. This could lead to significant storage consumption for each patched app by default. A checkbox labelled "Save a copy" should be included within the installation dialogue, accompanied by the subtitle: "A copy of the patched app will be saved, which can be installed or exported in the future. You can manage these copies in the 'Patched apps' section located in the main menu of ReVanced Manager." As suggested by @validcube, it is advisable to provide the option to delete saved APKs, preferably within the same interface where users can install, export, or manage the patched app. A submenu displaying the list of saved patched apps, their patching date, and metadata, including applied patches, should be available for that.

If the user checks the checkmark to save the patched app, but a saved patched app of the same version is already present, give the user the option to overwrite any existing saved app.

The installation button in the patched app view has to show the same popup as the one after patching, as you may want to mount the app.

**Comment by BenjaminHalko** [comment]
> This could lead to significant storage consumption for each patched app by default
I propose we only save the last 3 packages, plus there could be an option when you click on an app in the history to delete it.

> A checkbox labelled "Save a copy" should be included within the installation dialogue, accompanied by the subtitle: "A copy of the patched app will be saved, which can be installed or exported in the future. You can manage these copies in the 'Patched apps' section located in the main menu of ReVanced Manager."
How about just a toggle in the settings menu. I feel like the patch history should be used for the event where a user forgets to either export or install an app. There have been a few times where I have forgot to export my apk and then needed to repatch it. But never an event where I wanted the app to save it for me. If I needed it saved, I just exported the apk. If I had to click a checkbox every time I wanted the app to save the apk, that would defeat the point of having the app save a backup.

> The installation button in the patched app view has to show the same popup as the one after patching, as you may want to mount the app.
Already should do that, though I don't have root so I can't test that out.

**Comment by oSumAtrIX** [comment]
> I propose we only save the last 3 packages

How about using that as a maximum, but still inform the user about being able to overwrite the last saved patched app via a checkmark if a patched app is already saved? That way the user can replace their last saved patched app every time they successfully patch without having to manually clean up.

>  If I remembered to click a checkbox then I will have remembered to export the apk.

In this case, how about checking it by default. Having to visit the settings to toggle it, would be difficult right from the patch result screen.

**Comment by BenjaminHalko** [comment]
The "3 packages" limit would be 3 different packages. If you patch youtube it would replace youtube in the history automatically. If the youtube package was not in the history it would bump the oldest package off the history.

Although, thinking about it now, since the patch history is for "if the user forgets to preform an action once finished patching" it might be simpler to save ONLY the last patched apk. There isn't really a need to save more than that.

**Comment by oSumAtrIX** [comment]
> There isn't really a need to save more than that.

Yeah, 1 backup per app should be enough, saving the last patched app only (give the user the choice not to), should be good too.

**Comment by Ushie** [comment]
I'm closing this PR, the use case is very niche and the development time would better be spent in the compose-dev branch, I think that's better than spending the development and review time for a feature that will be short-lived

Very sorry, I know you were specifically assigned this, but the issue you were assigned was misunderstood at the time you were assigned to this

**Comment by Ushie** [comment]
This PR can be re-opened, if you choose to finish it

**Comment by oSumAtrIX** [comment]
Reopening as this PR is nearly finished.

**Comment by BenjaminHalko** [comment]
I'll try to finish it soon, just been busy with life lately. Sorry about that!

**Comment by BenjaminHalko** [comment]
All done!

**Comment by BenjaminHalko** [comment]
Now only 1 app is stored, you have the option to delete it by going into it's menu (same place where you would install/export, and you can now disable saving the app altogether in the settings.

**Comment by Ushie** [comment]
Requesting the review of osumatrix who's worked more closely with this, I can't thoroughly review the backend

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]
I suggest to consolidate the patched and installed apps sections to a "Patched apps" section. Installed patched apps list items can be distinct from not installed patched apps.

@PalmDevs Could you look into UI in this regard?

**Comment by PalmDevs** [comment]
@BenjaminHalko Can you upload a recording of what the feature looks like now? Or is the video in the PR description already up-to-date?

**Comment by BenjaminHalko** [comment]
Not shown in the video is the fact that only the last patched app gets saved. If there is already an app in the slot, it will get removed once a new app is done patching.

https://github.com/ReVanced/revanced-manager/assets/73490201/384eeec2-f954-40e3-a70f-237157e0ea1f

File: assets/i18n/en_US.json [hunk 1]
```diff
@@ -239,6 +241,9 @@
         "universalPatchesLabel": "Show universal patches",
         "universalPatchesHint": "Display all apps and universal patches (may slow down the app list)",
 
+        "patchHistoryLabel": "Save patched app",
+        "patchHistoryHint": "Save the last patched app to install or export later",
```
Comment (oSumAtrIX) [inline assets/i18n/en_US.json:245]:
```suggestion
        "patchHistoryHint": "Save the last patch to install or export later",
```

File: assets/i18n/en_US.json [hunk 2]
```diff
@@ -26,12 +26,14 @@
         "widgetTitle": "Dashboard",
 
         "updatesSubtitle": "Updates",
-        "patchedSubtitle": "Patched apps",
+        "patchHistorySubtitle": "Last patched app",
```
Comment (oSumAtrIX) [inline assets/i18n/en_US.json:29]:
```suggestion
        "patchHistorySubtitle": "Last patch",
```

**Review by oSumAtrIX — COMMENTED** [review]
I am not sure about this suggestion, but currently, what "Last patched app" means can be confusing to the user. Is it the now-installed app? If you press delete, will it delete my currently installed patched app? Maybe add text to the app view that explains, that it is a backup of the last patched application

**Comment by BenjaminHalko** [comment]
Added this description:
Screenshot_20240305-084404_ReVanced Manager Debug

**Comment by validcube** [comment]
Is there anything that's holding up this PR?

**Comment by BenjaminHalko** [comment]
I guess just approval.

**Comment by oSumAtrIX** [comment]
https://github.com/ReVanced/revanced-manager/pull/1414#pullrequestreview-1820135060

Has this been resolved?

**Comment by ILoveOpenSourceApplications** [comment]
> [#1414 (review)](https://github.com/ReVanced/revanced-manager/pull/1414#pullrequestreview-1820135060)
> 
> Has this been resolved?

@BenjaminHalko ?

**Comment by BenjaminHalko** [comment]
@ILoveOpenSourceApplications I'll check tmr.

I think some text just needs to be changed.

**Comment by oSumAtrIX** [comment]
Is it possible to display how much storage the last saved patched app takes?

File: android/settings.gradle [hunk 3]
```diff
@@ -18,7 +18,7 @@ pluginManagement {
 
 plugins {
     id "dev.flutter.flutter-plugin-loader" version "1.0.0"
-    id "com.android.application" version "8.4.1" apply false
+    id "com.android.application" version "8.2.0" apply false
```
Comment (oSumAtrIX) [inline android/settings.gradle:21]:
```suggestion
    id "com.android.application" version "8.4.1" apply false
```

File: assets/i18n/strings.i18n.json [hunk 4]
```diff
@@ -27,10 +27,12 @@
     "refreshSuccess": "Refreshed successfully",
     "widgetTitle": "Dashboard",
     "updatesSubtitle": "Updates",
-    "patchedSubtitle": "Patched apps",
+    "patchHistorySubtitle": "Last patched app",
```
Comment (oSumAtrIX) [inline assets/i18n/strings.i18n.json:30]: "history" should be renamed to lastPatchedApp everywhere

File: assets/i18n/strings.i18n.json [hunk 5]
```diff
@@ -256,18 +260,25 @@
...
     "unmountButton": "Unmount",
+    "exportButton": "Export",
+    "deleteButton": "Delete",
     "rootDialogTitle": "Error",
+    "lastPatchedAppDescription": "This is a backup of the app that was last patched.",
     "unmountDialogText": "Are you sure you want to unmount this app?",
     "uninstallDialogText": "Are you sure you want to uninstall this app?",
     "rootDialogText": "App was installed with superuser permissions, but currently ReVanced Manager has no permissions.\nPlease grant superuser permissions first.",
+    "removeAppDialogTitle": "Delete app?",
+    "removeAppDialogText": "Are you sure you want to delete this app?",
```
Comment (oSumAtrIX) [inline assets/i18n/strings.i18n.json:274]:
```suggestion
    "removeAppDialogText": "Are you sure you want to delete this backup?",
```

File: assets/i18n/strings.i18n.json [hunk 6]
```diff
@@ -256,18 +260,25 @@
...
     "packageNameLabel": "Package name",
     "installTypeLabel": "Installation type",
     "mountTypeLabel": "Mount",
     "regularTypeLabel": "Regular",
     "patchedDateLabel": "Patched date",
     "appliedPatchesLabel": "Applied patches",
+    "sizeLabel": "File size",
```
Comment (oSumAtrIX) [inline assets/i18n/strings.i18n.json:281]: Seems to be unused

**Comment by BenjaminHalko** [comment]
Screenshot_20240628-094156_ReVanced Manager Debug

File: assets/i18n/strings.i18n.json [hunk 7]
```diff
@@ -27,10 +27,12 @@
     "refreshSuccess": "Refreshed successfully",
     "widgetTitle": "Dashboard",
     "updatesSubtitle": "Updates",
-    "patchedSubtitle": "Patched apps",
+    "lastPatchedAppSubtitle": "Last patched app",
+    "patchedSubtitle": "Installed apps",
     "changeLaterSubtitle": "You can change this in the settings at a later time.",
     "noUpdates": "No updates available",
     "WIP": "Work in progress...",
+    "noHistory": "No app found",
```
Comment (oSumAtrIX) [inline assets/i18n/strings.i18n.json:35]: Should be called noSavedAppFound for example

File: lib/services/manager_api.dart [hunk 8]
```diff
@@ -685,6 +721,16 @@ class ManagerAPI {
     patchedApps.addAll(mountedApps);
 
     await setPatchedApps(patchedApps);
+
+    // Removed saved app history if the file is not found.
```
Comment (oSumAtrIX) [inline lib/services/manager_api.dart:725]:
```suggestion
    // Delete the saved app if the file is not found.
```

File: assets/i18n/strings.i18n.json (hunk 7 above)
Comment (oSumAtrIX) [inline assets/i18n/strings.i18n.json:35]:
```suggestion
    "noSavedAppFound": "No app found",
```

File: lib/ui/widgets/homeView/last_patched_app_card.dart [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+            child: Column(
+              children: <Widget>[
+                Icon(
+                  size: 40,
+                  Icons.update_disabled,
+                  color: Theme.of(context).colorScheme.secondary,
+                ),
+                const SizedBox(height: 16),
+                Text(
+                  t.homeView.noHistory,
```
Comment (oSumAtrIX) [inline lib/ui/widgets/homeView/last_patched_app_card.dart:28]:
```suggestion
                  t.homeView.noSavedAppFound,
```

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]

**Review by oSumAtrIX — APPROVED** [review]

**Pull request merged**